### PR TITLE
Checks if Sub and Ses ID in path and filename mismatch

### LINKED ISSUE #264: Report an error when sub or ses do not match between the filename and the folder name
For example: `/sub-s524/ses-1/anat/sub-s524_ses-2_T2w.nii.gz`

### PR DESCRIPTION
**Pull request by suyashdb**
closes #264.

### CODE REVIEW

**Comment by chrisgorgo** [comment]
Please resolve conflicts.

File: utils/issues/list.js [hunk 1]
```diff
@@ -306,5 +306,15 @@ module.exports = {
         key: 'QUICK_VALIDATION_FAILED',
         severity: 'error',
         reason: 'Quick validation failed - the general folder structure does not resemble a BIDS dataset. Have you chosen the right folder (with "sub-*/" subfolders)? Check for structural/naming issues and presence of at least one subject.',
+    },
+    62: {
+        key: 'sub-id in filename doesnt match directory',
+        severity: 'warning',
+        reason: 'sub-id in the filename doesn\'t match with the path of the file. File seems to be saved in incorrect subject directory.'
```
Comment (chrisgorgo) [inline utils/issues/list.js]: let's be consistent and call this sesssion label not `ses-id`

File: utils/issues/list.js [hunk 2]
```diff
@@ -306,5 +306,15 @@ module.exports = {
         key: 'QUICK_VALIDATION_FAILED',
         severity: 'error',
         reason: 'Quick validation failed - the general folder structure does not resemble a BIDS dataset. Have you chosen the right folder (with "sub-*/" subfolders)? Check for structural/naming issues and presence of at least one subject.',
+    },
+    62: {
+        key: 'sub-id in filename doesnt match directory',
+        severity: 'warning',
```
Comment (chrisgorgo) [inline utils/issues/list.js]: Why is this just a warning?

File: utils/issues/list.js [hunk 3]
```diff
@@ -306,5 +306,15 @@ module.exports = {
...
+        reason: 'sub-id in the filename doesn\'t match with the path of the file. File seems to be saved in incorrect subject directory.'
+    },
+    63: {
```
Comment (chrisgorgo) [inline utils/issues/list.js]: Is this used anywhere?

File: utils/issues/list.js [hunk 4]
```diff
@@ -308,13 +308,13 @@ module.exports = {
         reason: 'Quick validation failed - the general folder structure does not resemble a BIDS dataset. Have you chosen the right folder (with "sub-*/" subfolders)? Check for structural/naming issues and presence of at least one subject.',
     },
     62: {
-        key: 'SUBJECT_VALUE_CONTAINS_ILLEGAL_CHARECTER',
+        key: 'SUBJECT_LABEL_IN_FILENAME_DOESNOT_MATCH_DIRECTORY',
         severity: 'error',
-        reason: 'Sub label contain an Illegal Character hyphen or underscore. Please edit the filename as per BIDS spec.',
+        reason: 'subject label in the filename doesn\'t match with the path of the file. File seems to be saved in incorrect subject directory.'
```
Comment (chrisgorgo) [inline utils/issues/list.js]: capital "S" in "Subject"

File: utils/issues/list.js [hunk 5]
```diff
@@ -308,13 +308,13 @@ module.exports = {
         reason: 'Quick validation failed - the general folder structure does not resemble a BIDS dataset. Have you chosen the right folder (with "sub-*/" subfolders)? Check for structural/naming issues and presence of at least one subject.',
     },
     62: {
-        key: 'SUBJECT_VALUE_CONTAINS_ILLEGAL_CHARECTER',
```
Comment (chrisgorgo) [inline utils/issues/list.js]: Why are you modifying those errors instead of creating new ones?  Weren't the old ones already used?

**Comment by suyashdb** [comment]
Ready for review.

File: validators/bids.js [hunk 6]
```diff
@@ -432,6 +435,87 @@ BIDS = {
...
+            var pathValues = values[0];
+            var fileValues = values[1];
+            // console.log(path, '/n' ,values);
+
+            if (fileValues.sub !== null || fileValues.ses !== null) {
+                if (fileValues.sub !== pathValues.sub) {
+                    // console.log("before call  ",subses_mismatch);
+                    subses_mismatch = true;
+                    self.issues.push(new Issue({
+                        code: 61,
```
Comment (chrisgorgo) [inline validators/bids.js]: Not 64?

File: validators/bids.js [hunk 7]
```diff
@@ -432,6 +435,87 @@ BIDS = {
...
+                        evidence: "File: " + file.relativePath + " is saved in incorrect subject directory as per sub-id in filename.",
+                        file: file
+                    }));
+                }
+
+                if (fileValues.ses !== pathValues.ses) {
+                    subses_mismatch = true;
+                    self.issues.push(new Issue({
+                        code: 62,
```
Comment (chrisgorgo) [inline validators/bids.js]: Not 65?

File: tests/type.spec.js [hunk 8]
```diff
@@ -242,4 +243,43 @@ describe('utils.type.getPathValues', function () {
...
+             path: 'tests/data/BIDS-examples-1.0.0-rc3u5/ds001/sub-22/ses-1/func/sub-22_ses-2_task-rest_acq-prefrontal_physio.tsv.gz',
+             relativePath: '/sub-22/ses-1/func/sub-22_ses-2_task-rest_acq-prefrontal_physio.tsv.gz' },
+          '3':
+           { name: '/sub-25/ses-2/func/sub-22_ses-1_task-rest_acq-prefrontal_physio.tsv.gz',
+             path: 'tests/data/BIDS-examples-1.0.0-rc3u5/ds001/sub-25/ses-2/func/sub-22_ses-1_task-rest_acq-prefrontal_physio.tsv.gz',
+             relativePath: 'ds001//sub-25/ses-2/func/sub-22_ses-1_task-rest_acq-prefrontal_physio.tsv.gz' }},
+
+        callback = function (issues) {
+            for (var i in issues) {
+                if (issues[i]['code'] === 62) {
```
Comment (chrisgorgo) [inline tests/type.spec.js]: not 64?

File: tests/type.spec.js [hunk 9]
```diff
@@ -242,4 +243,43 @@ describe('utils.type.getPathValues', function () {
...
+                    code62_seen = true;
+                }
+                else if (issues[i]['code'] === 63) {
```
Comment (chrisgorgo) [inline tests/type.spec.js]: not 65?

File: tests/type.spec.js [hunk 10]
```diff
@@ -242,4 +243,43 @@ describe('utils.type.getPathValues', function () {
         assert.equal(utils.type.getPathValues('/sub-22/func/sub-22_task-rest_acq-prefrontal_physio.tsv.gz').sub, 22);
         assert.equal(utils.type.getPathValues('/sub-22/func/sub-22_task-rest_acq-prefrontal_physio.tsv.gz').ses, null);
     });
-});
\ No newline at end of file
+});
+
+describe('BIDS.subIDsesIDmismatchtest', function () {
+    // var issues = [{code: 61}, {code: 62}];
```
Comment (chrisgorgo) [inline tests/type.spec.js]: why is this commented out? can we remove this?

File: tests/type.spec.js [hunk 11]
```diff
@@ -242,4 +243,43 @@ describe('utils.type.getPathValues', function () {
...
+    var code62_seen = false;
+    var code63_seen = false;
```
Comment (chrisgorgo) [inline tests/type.spec.js]: the names of those variables do not correspond to the issue numbers anymore - this is confusing

**Review by chrisgorgo — CHANGES_REQUESTED** [review]
couple of small things, almost there!

**Pull request merged**